jobs/gtfs-rt-parser-v2/gtfs_rt_parser.py

```diff
@@ -0,0 +1,740 @@
+"""
+Parses binary RT feeds and writes them back to GCS as gzipped newline-delimited JSON
+"""
+import concurrent.futures
+import copy
+import datetime
+import gzip
+import hashlib
+import json
+import os
+import subprocess
+import tempfile
+import traceback
+from collections import defaultdict
+from concurrent.futures import ThreadPoolExecutor, Future
+from enum import Enum
+from functools import lru_cache
+from pathlib import Path
+from typing import ClassVar, Dict, List, Optional, Sequence, Tuple, Union, Any
+
+import backoff  # type: ignore
+import pendulum
+import typer
+from aiohttp.client_exceptions import (
+    ClientOSError,
+    ClientResponseError,
+    ServerDisconnectedError,
+)
+from calitp.storage import (
+    GTFSFeedType,
+    PartitionedGCSArtifact,
+    ProcessingOutcome,
+    fetch_all_in_partition,
+    get_fs,
+    JSONL_GZIP_EXTENSION,
+    GTFSRTFeedExtract,
+    GTFSFeedExtract,
+    AirtableGTFSDataExtract,
+    AirtableGTFSDataRecord,
+    get_latest_file,
+    PARTITIONED_ARTIFACT_METADATA_KEY,
+    GTFSScheduleFeedExtract,
+    make_name_bq_safe,
+)  # type: ignore
+from google.protobuf import json_format
+from google.protobuf.message import DecodeError
+from google.transit import gtfs_realtime_pb2  # type: ignore
+from pydantic import BaseModel, Field, validator, parse_obj_as
+from tqdm import tqdm
+
+RT_VALIDATOR_JAR_LOCATION_ENV_KEY = "GTFS_RT_VALIDATOR_JAR"
+JAR_DEFAULT = typer.Option(
+    os.environ.get(RT_VALIDATOR_JAR_LOCATION_ENV_KEY),
+    help="Path to the GTFS RT Validator JAR",
+)
+
+RT_PARSED_BUCKET = os.environ["CALITP_BUCKET__GTFS_RT_PARSED"]
+RT_VALIDATION_BUCKET = os.environ["CALITP_BUCKET__GTFS_RT_VALIDATION"]
+
+
+def make_dict_bq_safe(d: Dict[str, Any]) -> Dict[str, Any]:
+    return {
+        make_name_bq_safe(key): make_dict_bq_safe(value)
+        if isinstance(value, dict)
+        else value
+        for key, value in d.items()
+    }
+
+
+def serialize_pydantic_model_for_bigquery(model: BaseModel):
+    return make_dict_bq_safe(json.loads(model.json()))
+
+
+class RTProcessingStep(str, Enum):
+    parse = "parse"
+    validate = "validate"
+
+
+def log(*args, err=False, fg=None, pbar=None, **kwargs):
+    # capture fg so we don't pass it to pbar
+    if pbar:
+        pbar.write(*args, **kwargs)
+    else:
+        typer.secho(*args, err=err, fg=fg, **kwargs)
+
+
+def upload_if_records(
+    fs,
+    tmp_dir: str,
+    artifact: PartitionedGCSArtifact,
+    records: Sequence[Union[Dict, BaseModel]],
+    pbar=None,
+):
+    # BigQuery fails when trying to parse empty files, so shouldn't write them
```

Comment:
Agreed, I think we need a generic function to handle saving (or not saving) data. I've re-implemented this type of thing a couple times.

I will tinker in calitp-py, and it might be worth using in the unzipping PR